keps/sig-node/20190717-seccomp-ga.md

```diff
@@ -0,0 +1,371 @@
+    Seccomp  *SeccompOptions
+    ...
+}
+
+type SecurityContext struct {
+    ...
+    // The seccomp options to use by this container. If seccomp options are
+    // provided at both the pod & container level, the container options
+    // override the pod options.
+    // +optional
+    Seccomp  *SeccompOptions
+    ...
+}
+
+type SeccompOptions struct {
+    // The seccomp profile to run with.
+    SeccompProfile
+}
+
+// Only one profile source may be set.
+// +union
+type SeccompProfile struct {
+    // No seccomp profile should be set.
+    // +optional
+    Unconfined *bool
+    // Use a predefined profile defined by the runtime.
+    // Most runtimes only support "default"
+    // +optional
+    RuntimeProfile *string
+    // Load a profile defined in static file on the node.
+    // The profile must be preconfigured on the node to work.
+    // +optional
+    LocalhostProfile *string
+}
+```
+
+This API makes the options more explicit than the stringly-typed annotation values, and leaves room
+for new profile sources to be added in the future (e.g. Kubernetes predefined profiles or ConfigMap
+profiles). The seccomp options struct leaves room for future extensions, such as defining the
+behavior when a profile cannot be set.
+
+#### PodSecurityPolicy API
+
+```go
+type PodSecurityPolicySpec struct {
+    ...
+    // seccomp is the strategy that will dictate allowable and default seccomp
+    // profiles for the pod.
+    // +optional
+    Seccomp *SeccompStrategyOptions
+    ...
+}
+
+type SeccompStrategyOptions struct {
+    // The default profile to set on the pod, if none is specified.
+    // The default MUST be allowed by the allowedProfiles.
+    // +optional
+    DefaultProfile *v1.SeccompProfile
+
+    // The set of profiles that may be set on the pod or containers.
+    // If unspecified, seccomp profiles are unrestricted by this policy.
+    // +optional
+    AllowedProfiles *SeccompProfileSet
+}
+
+// A set of seccomp profiles. This struct should be a plural of v1.SeccompProfile.
+// All values are optional, and an unspecified field excludes all profiles of
+// that type from the set.
+type SeccompProfileSet struct {
+    // Whether the unconfined profile is included in this set.
+    // +optional
+    Unconfined *bool
+    // The allowed runtimeProfiles. A value of '*' allows all runtimeProfiles.
+    // +optional
+    RuntimeProfiles []string
+    // The allowed localhostProfiles. Values may end in '*' to include all
+    // localhostProfiles with a prefix.
+    // +optional
+    LocalhostProfiles []string
+}
+```
+
+## Design Details
+
+### Version Skew Strategy
+
+Because the API is currently represented as (mutable) annotations, care must be taken for migrating
+to the API fields. The cases to consider are: pod create, pod update, PSP create, PSP update.
+
+All API skew is resolved in the API server. New Kubelets will only use the seccomp values specified
+in the fields, and ignore the annotations.
+
+#### Pod Creation
```

Comment:
What is the expected behavior if the CRI runtime doesn't support seccomp? The pod should be created without seccomp (as in the current alpha annotation) or should result in an error?

Reply:
I think we need to maintain the backwards compatible behavior of ignoring it (I'll add a line to be explicit about this). Going forward, I left space in the API for a `FailureStrategy` which could dictate how to handle that situation (e.g. some applications may depend on seccomp application to run securely, while others might accept a best-effort best-practices approach).

Reply:
No application should depend on a policy; a policy can be applied as a non privileged operation by any code, so if it requires it then it can just apply it.